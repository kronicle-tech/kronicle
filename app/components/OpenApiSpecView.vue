<template>
  <div class="openapi-spec"></div>
</template>

<script lang="ts">
import Vue, { PropType } from 'vue'
import { MetaInfo } from 'vue-meta'
import {
  Component,
  OpenApiSpec,
  OpenApiSpecsState,
} from '~/types/kronicle-service'
import { findComponentState } from '~/src/componentStateUtils'

export default Vue.extend({
  props: {
    component: {
      type: Object as PropType<Component>,
      required: true,
    },
    openApiSpecIndex: {
      type: Number,
      required: true,
    },
  },
  head(): MetaInfo {
    return {
      script: [
        {
          src: 'https://cdn.jsdelivr.net/npm/redoc@2.0.0-rc.45/bundles/redoc.standalone.js',
        },
      ],
    }
  },
  mounted() {
    const openApiSpecs: OpenApiSpecsState | undefined = findComponentState(
      this.component,
      'openapi-specs'
    )
    this.load(openApiSpecs!.openApiSpecs[this.openApiSpecIndex - 1])
  },
  methods: {
    load(openApiSpec: OpenApiSpec) {
      const windowAny = window as any
      windowAny.Redoc.init(openApiSpec.spec, {}, this.$el)
    },
  },
})
</script>

<style scoped>
.openapi-spec {
  background-color: #fff;
}
</style>
