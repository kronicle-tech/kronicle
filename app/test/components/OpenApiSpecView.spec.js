import { createLocalVue, mount } from '@vue/test-utils'
import VueMeta from 'vue-meta'
import flushPromises from 'flush-promises'
import OpenApiSpecView from '~/components/OpenApiSpecView'

const localVue = createLocalVue()
localVue.use(VueMeta, { keyName: 'head' })

describe('OpenApiSpecView', () => {
  let propsData
  let wrapper

  async function createWrapper() {
    wrapper = await mount(OpenApiSpecView, {
      localVue,
      propsData
    })
  }

  beforeEach(() => {
    propsData = {
      component: {
        openApiSpecs: [
          {
            scannerId: 'test-scanner-1',
            url: 'https://example.com/test-1',
            description: 'Text Description 1',
            spec: {
              testField: 'test-value-1',
            },
          },
        ],
      },
      openApiSpecIndex: 1,
    }
    window.Redoc = { init: jest.fn() }
  })

  afterEach(() => {
    window.Redoc = undefined
    wrapper.destroy()
    wrapper = null
  })

  test('includes the script tag for Redoc', async () => {
    await createWrapper()
    expect(wrapper.vm.$metaInfo.script).toStrictEqual([
      {
        src: 'https://cdn.jsdelivr.net/npm/redoc@2.0.0-rc.45/bundles/redoc.standalone.js',
      }
    ])
  })

  test('renders the details of the correct OpenAPI spec', async () => {
    await createWrapper()
    expect(wrapper.element).toMatchSnapshot()
  })

  test('calls Redoc.init()', async () => {
    await createWrapper()
    await flushPromises()
    expect(window.Redoc.init).toBeCalledTimes(1)
    expect(window.Redoc.init.mock.calls[0][0]).toEqual(
      propsData.component.openApiSpecs[0].spec
    )
  })
})
