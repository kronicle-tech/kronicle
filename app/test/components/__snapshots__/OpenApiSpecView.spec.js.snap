// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenApiSpecView renders the details of the correct OpenAPI spec 1`] = `
<div
  class="openapi-spec"
/>
`;
